src/llm/wmt_model.py

```diff
@@ -247,20 +270,74 @@ def __init__(
                     model_dim=model_dim,
                     num_heads=num_heads,
                     expansion_dim=expansion_dim,
+                    dropout_rate=dropout_rate,
                 )
             )
 
     def forward(
         self,
-        decoder_input: Float[Tensor, "batch dec_seq_len model_dim"],
-        encoder_outputs: List[Float[Tensor, "batch enc_seq_len model_dim"]],
-    ) -> Float[Tensor, "batch dec_seq_len model_dim"]:
+        decoder_input: Float[Tensor, "batch dec_seq_len model_dim"],  # type: ignore
+        encoder_outputs: List[Float[Tensor, "batch enc_seq_len model_dim"]],  # type: ignore
+    ) -> Float[Tensor, "batch dec_seq_len model_dim"]:  # type: ignore
         x = decoder_input
         for idx, decoder_layer in enumerate(self.decoder_layers):
             x = decoder_layer(decoder_input=x, encoder_outputs=encoder_outputs[idx])
         return x
 
 
+class Embedding(nn.Module):
+    def __init__(self, model_dim: int, vocab_size: int):
+        super().__init__()
+        self.embedding = nn.Embedding(vocab_size, model_dim)
+        self.scale = math.sqrt(model_dim)
+
+    def forward(self, tokens: Float[Tensor, "batch seq_len"]) -> Float[Tensor, "batch seq_len model_dim"]:  # type: ignore
+        embedding = self.embedding(tokens) * self.scale
+        return embedding
+
+
+class PositionalEncoding(nn.Module):
+    def __init__(self, model_dim: int, max_seq_len: int, dropout_rate: float):
+        super().__init__()
+        self.position_encoding = torch.zeros(
+            max_seq_len, model_dim
+        )  # [max_seq_len, model_dim]
+        print(self.position_encoding.shape)
+        positions = torch.arange(0, max_seq_len).unsqueeze(1)  # [max_seq_len, 1]
+        print(positions.shape)
+        div_term = 1 / (
+            torch.pow(10000, torch.arange(0, model_dim, 2) / model_dim)
+        )  # [model_dim//2]
+        print(div_term.shape)
+        freq = positions * div_term  # [max_seq_len, model_dim//2]
+        print(freq.shape)
+        # all the positions but even spaced dimensions
+        self.position_encoding[:, 0::2] = torch.sin(freq)
+        self.position_encoding[:, 1::2] = torch.cos(freq)
+        self.residual_dropout = ResidualDropout(dropout_rate=dropout_rate)
+
+    def forward(
+        self, inputs: Float[Tensor, "batch seq_len model_dim"]  # type: ignore
+    ) -> Float[Tensor, "batch seq_len model_dim"]:  # type: ignore
+        x = inputs + self.position_encoding[: inputs.size(1)]
+        x = self.residual_dropout(x)
```

Comment:
I am not sure why we are doing this. Do we randomly dropout embeddings 👁️ ?

Reply:
I am doing because of my interpretation of this:

<img width="998" alt="image" src="https://github.com/user-attachments/assets/658c5fbf-43bb-4cb7-b918-f8e64f26a8bb" />

Reply:
Very interesting. Thanks for pointing me to the source.